pkg/service/auth_pipeline.go

```diff
@@ -498,19 +473,44 @@ func (pipeline *AuthPipeline) Evaluate() auth.AuthResult {
 			authResult <- result
 		}
 
-		metrics.ReportTimedMetric(authServerAuthConfigDurationMetric, evaluateFunc, pipeline.metricLabels()...)
+		metrics.ReportTimedMetric(metrics.Registry.GetAuthServerAuthConfigDurationMetric(), evaluateFunc, pipeline.metricLabels())
 	}()
 
 	return <-authResult
 }
 
 func (pipeline *AuthPipeline) reportStatusMetric(rpcStatusCode rpc.Code) {
-	metrics.ReportMetricWithStatus(authServerAuthConfigResponseStatusMetric, rpc.Code_name[int32(rpcStatusCode)], pipeline.metricLabels()...)
+	metrics.ReportMetricWithStatus(metrics.Registry.GetAuthServerAuthConfigResponseStatusMetric(), rpc.Code_name[int32(rpcStatusCode)], pipeline.metricLabels())
 }
 
-func (pipeline *AuthPipeline) metricLabels() []string {
-	labels := pipeline.AuthConfig.Labels
-	return []string{labels["namespace"], labels["name"]}
+func (pipeline *AuthPipeline) metricLabels() map[string]string {
+	labels := maps.Clone(pipeline.AuthConfig.Labels)
+	if labels == nil {
+		labels = make(map[string]string)
+	}
+
+	// Check for custom labels via the heuristic path
+	filteredMetadata := pipeline.GetRequest().GetAttributes().GetMetadataContext().GetFilterMetadata()
+	customLabels := filteredMetadata["io.kuadrant.metrics.labels"]
+	if customLabels != nil {
+		for k, v := range customLabels.Fields {
+			expr, err := cel.NewExpression(v.GetStringValue())
+			if err != nil {
+				pipeline.Logger.Error(err, "failed to parse custom label expression", "expression", v.String())
+				continue
+			}
+
+			if value, err := expr.ResolveFor(pipeline.GetAuthorizationJSON()); err != nil {
+				pipeline.Logger.Error(err, "failed to evaluate custom label expression", "expression", v.String())
+			} else if strValue, ok := value.(string); ok {
+				labels[k] = strValue
+			} else {
+				labels[k] = fmt.Sprintf("%v", value)
+			}
+		}
+	}
```

Comment:
I fear I may have changed that a bit this morning. See [these changes](https://github.com/Kuadrant/wasm-shim/pull/210/commits/47a1a8264787f5671f7e45f29640a648e69665cb?diff=unified&w=0), open to change it back tho. I sorta like that we don't conflate it all... but maybe that's a stupid idea... wdyt?